### PR TITLE
JSON ingest

### PR DESCRIPTION
**Pull request by eisber**
Goal: allow users to serialize C# objects to JSON (simple structure) using JSON.NET and 
- provide the object directly to VW and predict (w/o serializing to JSON and deserializing)
- provide the JSON serialized version to VW and learn (while still producing the same results as the direct version)

allows for direct ingestion of JSON into VW (without deserialization into intermediate objects).
allows for direct ingestion of C# objects optionally annotated using JSON.NET attributes.

based on JSON.NET - thus only supported through C#

Minor fixes: nuget build

### CODE REVIEW

File: cs/cs/Serializer/AnnotationJsonInspector.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                    || type == typeof(string)
+                    || (type.IsArray && IsNumericType(type.GetElementType()));
+        }
+
+        private static bool IsNumericType(Type type)
+        {
+            return type == typeof(double)
+                    || type == typeof(float)
+                    || type == typeof(decimal)
+                    || type == typeof(UInt16)
```
Comment (lhoang29) [inline cs/cs/Serializer/AnnotationJsonInspector.cs]: byte?

File: cs/cs/Serializer/PrefixedJsonReader.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    /// </summary>
+    internal class PrefixedJsonReader : JsonReader
+    {
+        private JsonReader reader;
+        private Stack<Tuple<JsonToken, object>> prefix;
+
+        internal PrefixedJsonReader(JsonReader reader, params Tuple<JsonToken, object>[] prefix)
+        {
+            this.reader = reader;
+            this.prefix = new Stack<Tuple<JsonToken, object>>(prefix.Reverse());
```
Comment (lhoang29) [inline cs/cs/Serializer/PrefixedJsonReader.cs]: can you use a queue here instead? (so no reverse)

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            using (VowpalWabbitMarshalContext defaultNamespaceContext = new VowpalWabbitMarshalContext(this.vw, context.ExampleBuilder))
+            {
+                if (!reader.Read())
+                    throw new ArgumentException("json");
+
+                if (reader.TokenType != JsonToken.StartObject)
+                    throw new ArgumentException("json");
+
+                Namespace defaultNamespace = new Namespace(this.vw);
+                using (defaultNamespaceContext.NamespaceBuilder = defaultNamespaceContext.ExampleBuilder.AddNamespace(' '))
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: maybe this '  ' space and the "_" characters below can be refactored out to const variables together somewhere for easy reference & documentation

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+            if (reader.TokenType != JsonToken.StartObject)
+                throw new ArgumentException("...");
+
+            if (!reader.Read())
+                throw new ArgumentException("...");
+
+            if (reader.TokenType != JsonToken.PropertyName)
+                throw new ArgumentException("...");
+
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: nit: you can potentially combine these 4 checks into a single statement or two separate ones i.e. if(!reader.Read() || tokenType != ...)

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                    case JsonToken.Integer:
+                        this.MarshalFloatFeature(context, ns, index, (float)(long)reader.Value);
+                        break;
+                    case JsonToken.Float:
+                        this.MarshalFloatFeature(context, ns, index, (float)(double)reader.Value);
+                        break;
+                    case JsonToken.EndArray:
+                        return;
+                    default:
+                        throw new ArgumentException("...");
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: is there something special with the ellipsis in exception message?

**Comment by lhoang29** [comment]
added some comments, looks good in general

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 3 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: done

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 5 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: Me being lazy while reporting errors.

Properly done.

File: cs/cs/Serializer/PrefixedJsonReader.cs (hunk 2 above)
Comment (eisber) [inline cs/cs/Serializer/PrefixedJsonReader.cs]: Better ☺ done.

Comment (JohnLangford) [inline cs/cs/Serializer/PrefixedJsonReader.cs]: Let's discuss tomorrow.

-John

File: cs/cs/Serializer/AnnotationJsonInspector.cs (hunk 1 above)
Comment (eisber) [inline cs/cs/Serializer/AnnotationJsonInspector.cs]: done

**Pull request merged**